File: bytes/bench/bench.ts
const L = process.argv[2] !== undefined ? parseInt(process.argv[2], 10) : 28;
const N = 1 << L;

function fillBuf(n: number): { b: Uint8Array; cs: number } {
  const b = new Uint8Array(n);
  for (let i = 0; i < n; i++) {
    b[i] = (i * 31 + 7) & 255;
  }
  b[n - 4] = 13;
  b[n - 3] = 10;
  b[n - 2] = 13;
  b[n - 1] = 10;
  const cs = b[12345] + b[n - 1];
  return { b, cs };
}

let t0 = performance.now();
const { b, cs: fillCs } = fillBuf(N);
console.log(`fill\t${performance.now() - t0}\t${fillCs}`);

t0 = performance.now();
let sum = 0 >>> 0;
for (let i = 0; i < N; i++) {
  sum = (sum + b[i]) >>> 0;
}
console.log(`sum\t${performance.now() - t0}\t${sum}`);

const needle = Buffer.from([13, 10, 13, 10]);
const buf = Buffer.from(b.buffer, b.byteOffset, b.byteLength);
t0 = performance.now();
const findIdx = buf.indexOf(needle);
console.log(`find\t${performance.now() - t0}\t${findIdx}`);

const sliceStart = (N / 4 + 1) | 0;
const sliceLen = (N / 2) | 0;
t0 = performance.now();
const s = Buffer.from(buf.subarray(sliceStart, sliceStart + sliceLen));
const sliceCs = s[0] + s[s.length - 1];
console.log(`slice\t${performance.now() - t0}\t${sliceCs}`);

t0 = performance.now();
const chunks: Buffer[] = [];
const numChunks = (N / 65536) | 0;
for (let k = 0; k < numChunks; k++) {
  chunks.push(Buffer.alloc(65536, k & 255));
}
const out = Buffer.concat(chunks);
if (out.length !== N) {
  process.exit(1);
}
const concatCs = out[0] + out[out.length - 1];
console.log(`concat\t${performance.now() - t0}\t${concatCs}`);

let x = 1 >>> 0;
let acc = 0 >>> 0;
const shift = 32 - L;
t0 = performance.now();
for (let i = 0; i < 1 << 24; i++) {
  x = (Math.imul(x, 1664525) + 1013904223) >>> 0;
  const idx = x >>> shift;
  acc = (acc + b[idx]) >>> 0;
}
console.log(`random\t${performance.now() - t0}\t${acc}`);

const c = Buffer.from(buf);
t0 = performance.now();
const eq = buf.equals(c);
console.log(`equal\t${performance.now() - t0}\t${eq ? 1 : 0}`);

for (const size of [1000000, 4000000]) {
  t0 = performance.now();
  let built = new Uint8Array(1);
  for (let i = 0; i < size; i++) {
    if (i === built.length) {
      const grown = new Uint8Array(built.length * 2);
      grown.set(built);
      built = grown;
    }
    built[i] = i & 255;
  }
  const buildCs = size + built[size - 1];
  console.log(`build_${size}\t${performance.now() - t0}\t${buildCs}`);
}

File: bytes/bench/bench.py
import sys
import time

L = int(sys.argv[1]) if len(sys.argv) > 1 else 28
N = 1 << L


def fill_buf(n: int) -> tuple[bytearray, int]:
    # Tile bytes(range(256)) then overwrite with the spec formula (same result as per-index fill).
    tile = bytes(range(256))
    reps = (n + 255) // 256
    b = bytearray((tile * reps)[:n])
    for i in range(n):
        b[i] = (i * 31 + 7) & 255
    b[n - 4 : n] = bytes([13, 10, 13, 10])
    cs = b[12345] + b[n - 1]
    return b, cs


t0 = time.perf_counter()
b, cs = fill_buf(N)
print(f"fill\t{(time.perf_counter() - t0) * 1000:.6f}\t{cs}")

t0 = time.perf_counter()
s = 0
for byte in b:
    s = (s + byte) & 0xFFFFFFFF
print(f"sum\t{(time.perf_counter() - t0) * 1000:.6f}\t{s}")

needle = bytes([13, 10, 13, 10])
t0 = time.perf_counter()
find_idx = b.find(needle)
print(f"find\t{(time.perf_counter() - t0) * 1000:.6f}\t{find_idx}")

slice_start = N // 4 + 1
slice_len = N // 2
t0 = time.perf_counter()
s_buf = bytearray(b[slice_start : slice_start + slice_len])
cs = s_buf[0] + s_buf[-1]
print(f"slice\t{(time.perf_counter() - t0) * 1000:.6f}\t{cs}")

t0 = time.perf_counter()
out = bytearray()
chunks = N // 65536
for k in range(chunks):
    out += bytes([k & 255]) * 65536
if len(out) != N:
    sys.exit(1)
cs = out[0] + out[-1]
print(f"concat\t{(time.perf_counter() - t0) * 1000:.6f}\t{cs}")

x, acc = 1, 0
shift = 32 - L
t0 = time.perf_counter()
for _ in range(1 << 24):
    x = (x * 1664525 + 1013904223) & 0xFFFFFFFF
    idx = x >> shift
    acc = (acc + b[idx]) & 0xFFFFFFFF
print(f"random\t{(time.perf_counter() - t0) * 1000:.6f}\t{acc}")

c = bytearray(b)
t0 = time.perf_counter()
eq = b == c
print(f"equal\t{(time.perf_counter() - t0) * 1000:.6f}\t{1 if eq else 0}")

for size in (1000000, 4000000):
    t0 = time.perf_counter()
    built = bytearray()
    for i in range(size):
        built.append(i & 255)
    cs = size + built[-1]
    print(f"build_{size}\t{(time.perf_counter() - t0) * 1000:.6f}\t{cs}")
